### PR TITLE
feat(sentryapp): Sentry apps request with dependentData

### PR DESCRIPTION
**Pull request by NisanthanNanthakumar**
## Objective:
This is the backend portion of the fix. Frontend is here: https://github.com/getsentry/sentry/pull/31881

This PR adds the `dependendData` query param for fields with a `depends_on` property. We also make the requests with the `dependentData` query param when loading a saved rule.

### CODE REVIEW

**Comment by mgaeta** [comment]
So the point of `get_attrs()` is to combine the N database calls you'd do for `item_list` into a single call. You should be doing 
```py
SentryAppInstallation.objects.get_installed_for_organization(
    kwargs["organization_id"]
)
.filter(sentry_app_id_in=sentry_app_ids)
.get()
```

File: src/sentry/api/serializers/models/alert_rule_trigger_action.py [hunk 1]
```diff
@@ -39,7 +42,28 @@ def get_input_channel_id(self, action):
         """
         return action.target_identifier if action.type == action.Type.SLACK.value else None
 
-    def serialize(self, obj, attrs, user):
+    def get_attrs(self, item_list, user, **kwargs):
```
Comment (mgaeta) [inline src/sentry/api/serializers/models/alert_rule_trigger_action.py:45]: In order to make the `SentryAppInstallation` lookup one call you'd need to do:
```py
def get_attrs(self, item_list, user, **kwargs):
    sentry_app_ids = {action.sentry_app_id for action in item_list}
    sentry_app_installations = SentryAppInstallation.objects.get_installed_for_organization(
        kwargs["organization_id"]
    ).filter(sentry_app_id_in=sentry_app_ids)
    sentry_app_installations_by_sentry_app_id = {
        installation.sentry_app_id: installation
        for installation in sentry_app_installations
    }

    result = defaultdict(dict)
    for action in item_list:
        sentry_app_installation = sentry_app_installations_by_sentry_app_id.get(action.sentry_app_id)
        if sentry_app_installation and action.sentry_app_config:
            component = sentry_app_installation.prepare_sentry_app_components(
                "alert-rule-action", None, action.sentry_app_config
            )
            result[action] = {"formFields": component.schema.get("settings", {})}

    return result
```
But each call to `prepare_sentry_app_components()` also does a DB lookup which is still bad.
You'll need to refactor that method to group those rows as well.

File: src/sentry/models/integrations/sentry_app_installation.py [hunk 2]
```diff
@@ -101,7 +101,12 @@ def prepare_sentry_app_components(self, component_type, project=None):
             return None
 
         try:
-            sentry_app_components.Preparer.run(component=component, install=self, project=project)
+            if values is None:
```
Comment (mgaeta) [inline src/sentry/models/integrations/sentry_app_installation.py:104]: Move this out of the `try` block.

File: src/sentry/api/serializers/models/alert_rule_trigger_action.py (hunk 1 above)
Comment (NisanthanNanthakumar) [reply src/sentry/api/serializers/models/alert_rule_trigger_action.py:45]: 👍  makes sense!

File: src/sentry/api/serializers/models/alert_rule_trigger.py [hunk 3]
```diff
@@ -58,7 +58,7 @@ def get_attrs(self, item_list, user, **kwargs):
             exclusions.append(project_slug)
         return result
 
-    def serialize(self, obj, attrs, user):
+    def serialize(self, obj, attrs, user, **kwargs):
         data = super().serialize(obj, attrs, user)
```
Comment (dashed) [inline src/sentry/api/serializers/models/alert_rule_trigger.py:62]: @NisanthanNanthakumar should `**kwargs` be passed to `super().serialize()` as well?

File: src/sentry/api/serializers/models/alert_rule_trigger_action.py [hunk 4]
```diff
@@ -39,7 +42,45 @@ def get_input_channel_id(self, action):
...
+        sentry_app_installations = SentryAppInstallation.objects.get_installed_for_organization(
+            kwargs["organization_id"]
+        ).filter(sentry_app_id__in=sentry_app_ids)
+        sentry_app_installations_by_sentry_app_id = {
+            installation.sentry_app_id: installation for installation in sentry_app_installations
+        }
+
+        components = SentryAppComponent.objects.filter(
+            sentry_app_id__in=sentry_app_ids, type="alert-rule-action"
+        )
```
Comment (markstory) [inline src/sentry/api/serializers/models/alert_rule_trigger_action.py:56]: Could these queries be combined using eager loading?

File: src/sentry/api/serializers/models/alert_rule_trigger.py (hunk 3 above)
Comment (NisanthanNanthakumar) [reply src/sentry/api/serializers/models/alert_rule_trigger.py:62]: 👍 good catch!

File: src/sentry/api/serializers/models/alert_rule_trigger_action.py (hunk 4 above)
Comment (NisanthanNanthakumar) [reply src/sentry/api/serializers/models/alert_rule_trigger_action.py]: @markstory ended up refactoring to move out all the SentryApp lookups to the endpoint and pass down as kwargs. I combined the queries using `.annotate` to get the related sentry_app_component for an installation.

**Comment by mgaeta** [comment]
The problem isn't that you're making DB calls in the serializer, it's valid and expected to make calls in `get_attrs()`. What we're trying to avoid is making O(N) DB calls serially if we want to serialize N AlertRuleTriggerActions. You should be able to move your queries back into the serializer and and leave the network calls in the endpoint.

**Comment by NisanthanNanthakumar** [comment]
> The problem isn't that you're making DB calls in the serializer, it's valid and expected to make calls in get_attrs(). What we're trying to avoid is making O(N) DB calls serially if we want to serialize N AlertRuleTriggerActions. You should be able to move your queries back into the serializer and and leave the network calls in the endpoint.

@mgaeta makes sense! ahh need to make those queries anyways for the SentryApp preparers, so I'll just keep as-is and pass them down.

File: src/sentry/api/serializers/models/alert_rule.py [hunk 5]
```diff
@@ -29,11 +36,32 @@ def get_attrs(self, item_list, user, **kwargs):
...
-        serialized_triggers = serialize(list(triggers))
+        serialized_triggers = serialize(list(triggers), **kwargs)
+
+        trigger_actions = AlertRuleTriggerAction.objects.filter(
+            alert_rule_trigger__alert_rule_id__in=alert_rules.keys()
+        ).exclude(sentry_app_config__isnull=True, sentry_app_id__isnull=True)
+
+        sentry_app_installations_by_sentry_app_id = (
+            SentryAppInstallation.objects.get_related_sentry_app_components(
+                organization_id__in={
```
Comment (mgaeta) [inline src/sentry/api/serializers/models/alert_rule.py:47]: This is not a `filter()` call. You'll need to pass `organization_ids` as a parameter.

File: src/sentry/api/endpoints/project_rule_details.py [hunk 6]
```diff
@@ -50,9 +52,32 @@ def get(self, request: Request, project, rule) -> Response:
...
+            rule,
+            request.user,
+        )
+
+        # Prepare Rule Actions that are SentryApp components using the meta fields
+        for action in serialized_rule.data.get("actions", []):
+            if action.get("_sentry_app_installation") and action.get("_sentry_app_component"):
+
+                component = SentryAppInstallation(
+                    **action.get("_sentry_app_installation", {})
```
Comment (mgaeta) [inline src/sentry/api/endpoints/project_rule_details.py:67]: Rehydrating the models from the serialized JSON might cause problem later if the constructor gets a parameter it wasn't prepared for.

File: src/sentry/api/serializers/models/alert_rule.py (hunk 5 above)
Comment (NisanthanNanthakumar) [reply src/sentry/api/serializers/models/alert_rule.py:47]: good catch!

File: src/sentry/api/endpoints/project_rule_details.py (hunk 6 above)
Comment (NisanthanNanthakumar) [reply src/sentry/api/endpoints/project_rule_details.py:67]: ehh it seems low risk bc the JSON is from the model originally. Do you have a suggestion on a better way?

File: src/sentry/api/serializers/models/rule.py [hunk 7]
```diff
@@ -56,6 +57,26 @@ def get_attrs(self, item_list, user, **kwargs):
         rules = {item.id: item for item in item_list}
         resolved_actors = {}
         owners_by_type = defaultdict(list)
+
+        sentry_app_uuids = {
```
Comment (mgaeta) [inline src/sentry/api/serializers/models/rule.py:61]:
```py
sentry_app_uuids = {
  action.get("sentryAppInstallationUuid") 
  for rule in rules.values() 
  for action in rule.get("actions", [])
}

Comment (NisanthanNanthakumar) [reply src/sentry/api/serializers/models/rule.py:61]: good catch!

**Review by mgaeta — APPROVED** [review]

**Pull request merged**